### PR TITLE
#1369: added latest stable version pattern

### LINKED ISSUE #1369: Ability to select latest stable version via version pattern.
As a devonfw-ide user, I usually want to get stable versions of tools so that I do not end up with bugs and being early adopter.

In devonfw-ide we have support for version prefixes (see #893 and #1112).
However, in IDEasy, we have implemented a more advanced version pattern matching that can distinguish between "latest stable version" and "latest version including unstable".
We also need to implement this feature in devonfw-ide.
Otherwise adding latest alpha/beta versions to ide-urls cannot be done (as then projects will end up getting these alpha versions by accident) and users are blocked from using such versions at all even if they explicitly want or need this. See https://github.com/devonfw/IDEasy/issues/138 as an example.

### PR DESCRIPTION
**Pull request by MattesMrzik**
Closes #1369. To select the latest version (even unstable), one can pass \*! or something like 1.2*!

### CODE REVIEW

**Comment by CREITZ25** [comment]
@MattesMrzik Looks good, but check variables between line 1914 ans 1930. 
For example "$version" should be written with brackets "${version}", or not?

**Review by hohwille — CHANGES_REQUESTED** [review]
@MattesMrzik thanks for this great PR. You also cleanup the code removing lots of typos and trailing spaces. Your implementation to get the latest stable version also looks solid and has nice documentation.
As already commented, you are using a tool that is not available on MacOS so you need to adapt that part and find a different solution.
Further you missed a small but important detail from the story demand:
You need to invert the invocations of the functions `getLatestSoftwareVersion` and `getLatestStableSoftwareVersion`.
See from IDEasy:
```
  /** {@link VersionIdentifier} "*" that will resolve to the latest stable version. */
  public static final VersionIdentifier LATEST = VersionIdentifier.of("*");
```

If the version is undefined we also want to install the latest *stable* version, what is the main purpose of the story as devonfw-ide users got unstable versions installed by default if those are the latest ones available.

**Comment by hohwille** [comment]
setup test failed with:
> ******** ATTENTION ********
Could not resolve version prefix 17* : no matching version found in /home/runner/work/ide/ide/scripts/target/integration-test/test-setup/urls/java/java/
We are sorry for the inconvenience. Please check the above errors, resolve them and try again.
Command 'java setup silent' failed with exit code 255

File: scripts/src/main/resources/scripts/functions [hunk 1]
```diff
@@ -1857,7 +1869,69 @@ function doGetLatestSoftwareVersion() {
...
+# $1: software
+# $2: optional prefix
+function doGetLatestStableSoftwareVersion() {
+  local software="${1}"
+  local prefix="${2}"
+  local edition
+  local version
+  if [ -z "${prefix}" ]
+  then
+    mapfile -t versions < <(find "${DEVON_IDE_HOME}"/urls/"${software}"/"${edition}" -mindepth 1 -maxdepth 1 -print | awk -F'/' '{print $NF}' | sort -rV)
```
[the remark this answers is not in the document]
Comment (MattesMrzik) [reply scripts/src/main/resources/scripts/functions]: I changed it to make it compatible with macOS

**Review by hohwille — APPROVED** [review]
@MattesMrzik thanks for your fix. Now complete and ready for merge. Good job 👍

**Pull request merged**